main/lsp/LSPTask.h

```diff
@@ -35,14 +41,51 @@ class LSPTask {
                                         std::vector<core::FileRef> frefs) const;
     LSPQueryResult queryBySymbol(LSPTypecheckerDelegate &typechecker, core::SymbolRef symbol) const;
 
-    LSPTask(const LSPConfiguration &config, bool enableMultithreading);
+    LSPTask(const LSPConfiguration &config, LSPMethod method);
 
 public:
-    const bool enableMultithreading;
+    virtual ~LSPTask();
+
+    const LSPMethod method;
+
+    // Measures the end-to-end latency of this task from the first message received from the client. Can be nullptr
+    // if not relevant.
+    std::unique_ptr<Timer> latencyTimer;
+
+    enum class Phase {
+        PREPROCESS = 1,
+        INDEX = 2,
+        RUN = 3,
+    };
+
+    /**
+     * If `true`, this task can be delayed in favor of processing other tasks sooner. Defaults to `false`.
+     */
+    virtual bool isDelayable() const;
+
+    // Attempts to cancel the task if it corresponds to the given request ID. Returns `true` if cancelation succeeds.
+    // The default implementation returns `false`.
+    virtual bool cancel(const MessageId &id);
+
+    virtual bool canPreempt(const LSPIndexer &) const;
 
-    virtual ~LSPTask() = default;
+    virtual bool needsMultithreading(const LSPIndexer &) const;
 
-    // Runs the task. Is only ever invoked from the typechecker thread.
+    // Returns the phase at which the task is complete. Some tasks only need to interface with the preprocessor or the
+    // indexer. The default implementation returns RUN.
+    virtual Phase finalPhase() const;
+
+    // Some tasks, like request cancelations, need to interface with the preprocessor. The default implementation is
+    // a no-op. Is only ever invoked from the preprocessor thread.
+    virtual void preprocess(LSPPreprocessor &preprocessor);
```

Comment:
These three methods are at the heart of task processing.

Tasks are preprocessed, indexed, and then run. Most tasks only need to implement `run`, but a few special tasks interact with the indexer and preprocessor. I anticipate that, in the future, we may see more simple tasks that don't need to interact with the typechecker at all.